src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs

```diff
@@ -262,11 +265,15 @@ private void UpdateReferenceLocationsTask(Task<IInlineRenameLocationSet> allRena
 
             var asyncToken = _asyncListener.BeginAsyncOperation("UpdateReferencesTask");
             _allRenameLocationsTask = allRenameLocationsTask;
-            allRenameLocationsTask.SafeContinueWith(
-                t => RaiseSessionSpansUpdated(t.Result.Locations.ToImmutableArray()),
+            allRenameLocationsTask.SafeContinueWithFromAsync(
+                async t =>
+                {
+                    await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, _cancellationTokenSource.Token);
+                    RaiseSessionSpansUpdated(t.Result.Locations.ToImmutableArray());
+                },
                 _cancellationTokenSource.Token,
-                TaskContinuationOptions.OnlyOnRanToCompletion,
-                ForegroundTaskScheduler).CompletesAsyncOperation(asyncToken);
+                TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
```

Comment:
what does it mean to call `SwitchToMainThread(alwaysYield: true` with `ExecuteSynchronously`?

Reply:
In this comment, I'll explain what this code is doing in its new form. With that knowledge, I'm open to suggestions regarding the approach we take moving forward.

The new code is an *ugly* pattern, but not a *fragile* pattern. It's also used in several locations in this pull request, each time with the same intent. The goal of the change is to execute a continuation on the main thread, which is roughly equivalent to the following:

```csharp
await ExecuteSomeOperationOnBackgroundAsync(cancellationToken);
await SwitchToMainThreadAsync(cancellationToken);
SomeForegroundOperation();
```

While the old code used a special `TaskScheduler` to switch to the main thread, the new code uses `SwitchToMainThreadAsync`. The `ExecuteSynchronously` involves the code that actually makes the call to `SwitchToMainThreadAsync` - it turns out this code can run anywhere, so I use `ExecuteSynchronously` to avoid a situation where the continuation is scheduled on a background thread only to switch right back to the main thread.

For clarity, it make sense to (at some point) update the code to use async/await rather than the `ContinueWith` pattern seen today. However, as mentioned above this change would be a clarity/readability change only and not something truly necessary. We could create a tracking issue for updating these, but in the scheme of things it seems relatively low priority. Perhaps a more interesting change would be creating an analyzer/fixer as part of Microsoft/vs-threading which helps code using this form as part of the general process of moving from alternatives to vs-threading.